pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java

```diff
@@ -0,0 +1,195 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.realtime.impl.dictionary;
+
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.pinot.core.io.readerwriter.PinotDataBufferMemoryManager;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+
+
+/**
+ * Off-heap variable length mutable bytes store.
+ * <p>The class is thread-safe for single writer and multiple readers.
+ * <p>There are two sets of buffers allocated for the store, one set to store the value end offsets and the other set to
+ * store the values. The new buffer is allocated when the previous buffer is filled up. The first entry of each offset
+ * buffer stores the end offset of the last value (or 0 for the first offset buffer) so that we can always get the
+ * previous and current value's end offset from the same buffer.
+ */
+public class OffHeapMutableBytesStore implements Closeable {
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  // MAX_NUM_BUFFERS = 8_192
+  private static final int MAX_NUM_BUFFERS = 1 << 13;
+
+  // OFFSET_BUFFER_SIZE = 32_768 + 4
+  private static final int OFFSET_BUFFER_SHIFT_OFFSET = 13;
+  private static final int OFFSET_BUFFER_SIZE = ((1 << OFFSET_BUFFER_SHIFT_OFFSET) + 1) << 2;
+  private static final int OFFSET_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - OFFSET_BUFFER_SHIFT_OFFSET);
+
+  // VALUE_BUFFER_SIZE = 1_048_576
+  private static final int VALUE_BUFFER_SIFT_OFFSET = 20;
+  private static final int VALUE_BUFFER_SIZE = 1 << VALUE_BUFFER_SIFT_OFFSET;
+  private static final int VALUE_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - VALUE_BUFFER_SIFT_OFFSET);
+
+  private final PinotDataBuffer[] _offsetBuffers = new PinotDataBuffer[MAX_NUM_BUFFERS];
+  private final PinotDataBuffer[] _valueBuffers = new PinotDataBuffer[MAX_NUM_BUFFERS];
+  private final PinotDataBufferMemoryManager _memoryManager;
+  private final String _allocationContext;
+
+  private int _numValues;
+  private int _valueStartOffset;
```

Comment:
Please add a comment on how this value is computed. If I understand right, it is the GLOBAL start offset of each value assuming a linear array of (VALUE_ BUFFER_SIZE * numBuffers). If a value cannot be held in one buffer, then the _valueStartOffset is set to the start of the next buffer.

Reply:
Added